archive/imo/imo1972_b2.lean

```diff
@@ -0,0 +1,127 @@
+/-
+Copyright (c) 2020 Ruben Van de Velde, Stanislas Polu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ruben Van de Velde, Stanislas Polu
+-/
+
+import data.real.basic
+import data.set.basic
+import analysis.normed_space.basic
+
+/-- IMO 1972 B2
+
+Problem: `f` and `g` are real-valued functions defined on the real line. For all `x` and `y`,
+`f(x + y) + f(x - y) = 2f(x)g(y)`. `f` is not identically zero and `|f(x)| ≤ 1` for all `x`.
+Prove that `|g(x)| ≤ 1` for all `x`.
+
+This proof consists in introducting the supremum of `f`, `k ≤ 1` as well as `k' = k / ∥g y∥`. We then
+suppose that the conclusion does not hold (`hneg`) and show that `k ≤ k'` (by
+`2 * (∥f x∥ * ∥g y∥) ≤ 2 * k` obtained from the main hypothesis `hf1`) and  that `k' < k` (obtained
+from `hneg` directly), finally raising a contradiction with `k' < k'`.
+
+(Authored by Stanislas Polu inspired by Ruben Van de Velde).
+-/
+example (f g : ℝ → ℝ)
+  (hf1 : ∀ x, ∀ y, (f(x+y) + f(x-y)) = 2 * f(x) * g(y))
+  (hf2 : ∀ y, ∥f(y)∥ ≤ 1)
+  (hf3 : ∃ x, f(x) ≠ 0)
+  (y : ℝ) :
+  ∥g(y)∥ ≤ 1 :=
+begin
+  classical,
+  set S := set.range (λ x, ∥f x∥),
+  -- Introduce `k` the supremum of `f`.
+  let k : ℝ := Sup (S),
+
+  -- Show that `∥f x∥ ≤ k`.
+  have hk₁ : ∀ x, ∥f x∥ ≤ k,
+  { have h : ∃ x, ∀ y ∈ S, y ≤ x,
+    { use 1,
+      intros _ hy,
+      rw set.mem_range at hy,
+      rcases hy with ⟨z, rfl⟩,
+      exact hf2 z, },
+    intro x,
+    exact real.le_Sup S h (set.mem_range_self x), },
+  -- Show that `2 * (∥f x∥ * ∥g y∥) ≤ 2 * k`.
+  have hk₂ : ∀ x, _,
```

Comment:
Could you write the type explicitly here? I know it's inferred from the calc block below. But this is so close, yet so far, to being something legible.

Reply:
Done 👍